addons/account/models/company.py

```diff
@@ -232,10 +234,79 @@ def _validate_fiscalyear_lock(self, values):
+            if country_code == 'CA':
+                module_list.append('l10n_ca_check_printing')
+            if country_code in ['US', 'AU', 'NZ', 'CA', 'CO', 'EC', 'ES', 'FR', 'IN', 'MX', 'GB']:
+                module_list.append('account_yodlee')
+            if country_code in SYSCOHADA_LIST + [
+                'AT', 'BE', 'CA', 'CO', 'DE', 'EC', 'ES', 'ET', 'FR', 'GR', 'IT', 'LU', 'MX', 'NL', 'NO',
+                'PL', 'PT', 'RO', 'SI', 'TR', 'GB', 'VE', 'VN'
+                ]:
+                module_list.append('base_vat')
+            if country_code == 'MX':
+                module_list.append('l10n_mx_edi')
+
+            # European countries will be using SEPA
+            europe = self.env.ref('base.europe', raise_if_not_found=False)
+            if europe:
+                europe_country_codes = [x.code for x in europe.country_ids]
+                if country_code in europe_country_codes:
+                    module_list.append('account_sepa')
+                    module_list.append('account_bank_statement_import_camt')
+            module_ids = self.env['ir.module.module'].search([('name', 'in', module_list), ('state', '=', 'uninstalled')])
+            #module is being installed.
+            #when module installation is in process then button_install work perfect but 
+            #when creating or updating a company at that time we need call button_immediate_install because no installation
+            if(from_init):
+                module_ids.sudo().button_install()
+            else:
+                #a new company is being created.
+                self.env.user.company_ids += self
+                #localization for the given country isn't installed.
+                if module_ids:
+                    #all localization module call function try_loading from data file so pass company in context so COA is load of this company.
+                    module_ids.sudo().with_context({'allowed_company_ids': [self.id]}).button_immediate_install()
+                #filter out the list of templates that belong to newly created company's country and load the first one.
+                chart_template_xml_ids = self.env['ir.model.data'].search([('module', 'in', module_list), ('model', '=', 'account.chart.template')], limit=1)
+                if chart_template_xml_ids and not self.chart_template_id:
+                    chart_template = self.env['account.chart.template'].browse(chart_template_xml_ids.res_id)
+                    chart_template.try_loading(company=self)
+
+    @api.model
+    def create(self, vals):
+        company = super(ResCompany, self).create(vals)
+        company.sudo()._install_localization_packages()
+        return company
+
     def write(self, values):
         #restrict the closing of FY if there are still unposted entries
         self._validate_fiscalyear_lock(values)
-
+        company_without_coa = self.filtered(lambda c: not c.chart_template_id)
```

Comment:
Better to use "existing accounting" method, but better to check it only when not self.country_id and values.get('country_id')  I don't think you check not self.country_id here

Reply:
@jva-odoo